### PR TITLE
Language version parsing for Nextflow

### PR DESCRIPTION
**Pull request by coverbeck**
**Description**
Adds language version parsing for Nextflow. See [this comment](https://github.com/dockstore/dockstore/issues/5268#issuecomment-1377995755) for some details, but in short:

* The Nextflow DSL version can be specified in the nextflow.config and/or the main script file; if it is set, we set the version in Dockstore accordingly.
* If set in both the config and the script file, the script file wins.
* If it's not set in the nextflow.config or main script file, we don't set the version in Dockstore -- in that case the version is determined by the engine version or an environment variable setting, and we can't assume any of that.

**Review Instructions**
Update a Nextflow workflow (the nf-core workflows all have the DSL version explicitly specified, if you want to fork one of those); you should see the version appear in the UI on the versions page.

**Issue**
#5268

**Security**
If there are any concerns that require extra attention from the security team, highlight them here.

Please make sure that you've checked the following before submitting your pull request. Thanks!

- [x] Check that you pass the basic style checks and unit tests by running `mvn clean install`
- [x] Ensure that the PR targets the correct branch. Check the milestone or fix version of the ticket.
- [x] Follow the existing JPA patterns for queries, using named parameters, to avoid SQL injection
- [x] If you are changing dependencies, check the Snyk status check or the dashboard to ensure you are not introducing new high/critical vulnerabilities
- [x] Assume that inputs to the API can be malicious, and sanitize and/or check for Denial of Service type values, e.g., massive sizes
- [x] Do not serve user-uploaded binary images through the Dockstore API
- [x] Ensure that endpoints that only allow privileged access enforce that with the `@RolesAllowed` annotation
- [x] Do not create cookies, although this may change in the future
- [x] If this PR is for a user-facing feature, create and link a documentation ticket for this feature (usually in the same milestone as the linked issue). Style points if you create a documentation PR directly and link that instead.

### CODE REVIEW

**Comment by coverbeck** [comment]
Tests are failing due to GitHub rate limits, putting this into review in hopes of getting those resolved later today (and hopefully there are no real failures hidden behind those GitHub failures).

**Review by denis-yuen — APPROVED** [review]

**Comment by denis-yuen** [comment]
(I may have some ideas for improving the caching behaviour, stay tuned)

File: dockstore-webservice/src/main/java/io/dockstore/webservice/languages/NextflowHandler.java [hunk 1]
```diff
@@ -102,12 +103,28 @@ public Version parseWorkflowContent(String filepath, String content, Set<SourceF
...
+     *  based on the source files's descriptor type version.
+     * @param sourceFiles
+     * @param version
+     * @param configuration
+     * @param mainScript
+     */
+    private void updateDescriptorTypeVersion(final Set<SourceFile> sourceFiles, final Version version,
+            final Configuration configuration, final Optional<SourceFile> mainScript) {
+        final String dslVersion = this.calculateDslVersion(configuration, mainScript).orElse(null);
+        sourceFiles.forEach(sourceFile -> sourceFile.setTypeVersion(dslVersion));
```
Comment (kathy-t) [inline dockstore-webservice/src/main/java/io/dockstore/webservice/languages/NextflowHandler.java:124]: Does this set the version for all source files, including non-descriptor files like test parameter files? I think we only want to set it for descriptor files

Comment (coverbeck) [reply dockstore-webservice/src/main/java/io/dockstore/webservice/languages/NextflowHandler.java:124]: Ah, good catch!

Comment (denis-yuen) [reply dockstore-webservice/src/main/java/io/dockstore/webservice/languages/NextflowHandler.java]: (I vaguely recall someone from nf-core mentioning that they were working on it)

File: dockstore-webservice/src/main/java/io/dockstore/webservice/languages/NextflowHandler.java [hunk 2]
```diff
@@ -623,6 +644,70 @@ protected Map<String, DockerParameter> getCallsToDockerMap(String mainDescriptor
...
+     *     <li>If it's in neither, we can't calculate the version -- it's based on on the engine version
+     *     or a environment setting</li>
+     * </ol>
+     * @param configuration
+     * @param mainScript
+     * @return
+     */
+    private Optional<String> calculateDslVersion(final Configuration configuration, final Optional<SourceFile> mainScript) {
+        final Optional<String> dslInScript = mainScript.flatMap(sf -> getDslVersion(sf.getContent()));
+        if (dslInScript.isEmpty()) {
```
Comment (svonworl) [inline dockstore-webservice/src/main/java/io/dockstore/webservice/languages/NextflowHandler.java:699]: No need to change, but if we wanted to go full-bore on the streaminess: https://docs.oracle.com/en/java/javase/17/docs/api/java.base/java/util/Optional.html#or(java.util.function.Supplier)
The resulting "one-liner" is something like:
```
return mainScript
    .flatMap(sf -> getDslVersion(sf.getContent()))
    .or(() -> Optional.ofNullable(configuration.getString("nextflow.enable.dsl", null)));
```

**Review by svonworl — APPROVED** [review]

File: dockstore-webservice/src/main/java/io/dockstore/webservice/languages/NextflowHandler.java (hunk 2 above)
Comment (coverbeck) [reply dockstore-webservice/src/main/java/io/dockstore/webservice/languages/NextflowHandler.java:699]: Cool, I didn't actually know about `or`, and I do prefer your suggestion.

**Review by kathy-t — APPROVED** [review]

**Pull request merged**